File: autoviz/examples/python/09_tutorial_plot.py
#!/usr/bin/env python3
"""Publish numeric series for Autoviz Plot panel.

  /fake/plot/sine   std_msgs/Float64   → field data
  /fake/plot/cosine std_msgs/Float64   → field data
  /fake/plot/cmd    geometry_msgs/Twist → linear.x / angular.z

In Autoviz Plot: Add series, channel + field_path as above.

  /usr/bin/python3 examples/python/09_tutorial_plot.py
  ./build/autonomy/bin/autoviz
"""

from __future__ import annotations

from _reexec import early_reexec_if_needed

early_reexec_if_needed()

import argparse
import math
import time

from _bootstrap import prepare_example_environment

prepare_example_environment()

import autolink
from automsgs.msgs.geometry_msgs import twist_pb2
from automsgs.msgs.std_msgs import float64_pb2


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument('--rate', type=float, default=50.0)
    args = ap.parse_args()

    autolink.init('plot')
    n = autolink.Node('/autoviz/plot')
    w_sin = n.create_writer('/fake/plot/sine', float64_pb2.Float64, qos_depth=10)
    w_cos = n.create_writer('/fake/plot/cosine', float64_pb2.Float64, qos_depth=10)
    w_cmd = n.create_writer('/fake/plot/cmd', twist_pb2.Twist, qos_depth=10)
    rate = autolink.Rate(args.rate)
    print(f'plot @ {args.rate} Hz → /fake/plot/{{sine,cosine,cmd}}')

    t0 = time.time()
    try:
        while not autolink.is_shutdown():
            t = time.time() - t0
            w_sin.write(float64_pb2.Float64(data=math.sin(t)))
            w_cos.write(float64_pb2.Float64(data=math.cos(t * 1.5)))
            cmd = twist_pb2.Twist()
            cmd.linear.x = 0.5 + 0.5 * math.sin(t * 0.7)
            cmd.angular.z = 0.3 * math.cos(t)
            w_cmd.write(cmd)
            rate.sleep()
    except KeyboardInterrupt:
        pass
    finally:
        autolink.shutdown()


if __name__ == '__main__':
    main()

File: autoviz/examples/cpp/09_tutorial_plot.cpp
/******************************************************************************
 * Copyright 2026 The Openbot Authors (duyongquan)
 *****************************************************************************/

#include <cmath>
#include <iostream>
#include <memory>

#include <automsgs/msgs/geometry_msgs/twist.pb.h>
#include <automsgs/msgs/std_msgs/float64.pb.h>

#include "common/tutorial_utils.hpp"

using autoviz::examples::MakeWriter;
using autoviz::examples::NowSec;
using autoviz::examples::ParseRate;

int main(int argc, char** argv) {
  const double rate_hz = ParseRate(argc, argv, 50.0);

  if (!autolink::Init(argv[0])) return 1;
  auto node = autolink::CreateNode("/autoviz/plot");
  auto w_sin =
      MakeWriter<automsgs::msgs::std_msgs::Float64>(node, "/fake/plot/sine");
  auto w_cos =
      MakeWriter<automsgs::msgs::std_msgs::Float64>(node, "/fake/plot/cosine");
  auto w_cmd =
      MakeWriter<automsgs::msgs::geometry_msgs::Twist>(node, "/fake/plot/cmd");

  autolink::Rate rate(rate_hz);
  std::cout << "plot @ " << rate_hz
            << " Hz → /fake/plot/{sine,cosine,cmd}\n";

  const double t0 = NowSec();
  while (autolink::OK()) {
    const double t = NowSec() - t0;

    auto sin_msg = std::make_shared<automsgs::msgs::std_msgs::Float64>();
    sin_msg->set_data(std::sin(t));
    w_sin->Write(sin_msg);

    auto cos_msg = std::make_shared<automsgs::msgs::std_msgs::Float64>();
    cos_msg->set_data(std::cos(t * 1.5));
    w_cos->Write(cos_msg);

    auto cmd = std::make_shared<automsgs::msgs::geometry_msgs::Twist>();
    cmd->mutable_linear()->set_x(0.5 + 0.5 * std::sin(t * 0.7));
    cmd->mutable_angular()->set_z(0.3 * std::cos(t));
    w_cmd->Write(cmd);

    rate.Sleep();
  }
  return 0;
}
